Deprecation and Release Policy. (#3238)
ONSdigital/design-system: policies/deprecation-release-policy.md

## Release Policy

### Versioning

We follow [Semantic Versioning](https://semver.org/) to manage releases:

-   **Major Releases (X.0.0)**: Introduce breaking changes. Deprecated items are removed, and significant new features or changes are added.
-   **Minor Releases (X.Y.0)**: Add new features and enhancements in a backward-compatible manner. May include deprecations.
-   **Patch Releases (X.Y.Z)**: Include backward-compatible bug fixes and documentation updates.

### Release Schedule

-   **Breaking Change Releases**:
    -   Breaking change releases will take place every 6 months.

### Release Process

1. **Pre-release**:

    - **Feature Freeze**: Before a major or minor release, a feature freeze period is implemented to ensure stability.
    - **Testing**: Comprehensive testing is conducted, including unit tests, integration tests, and user acceptance tests.

2. **Release**:

    - **Release Notes**: Detailed release notes are published with each release, documenting new features, bug fixes, deprecations, and breaking changes.

>     - **Versioning**: The new version is tagged in the repository and published, and the library is published to the package manager (e.g. npm).

Reviewer: I was thinking this should probably be just this, npm isnt an example of a package manager we publish to, its the actual one
```suggestion
    - **Versioning**: The new version is tagged in the repository and published, and the library is published to the package manager npm.
```

>     - **Change Log**: Our change log is available on [GitHub](https://github.com/ONSdigital/design-system/releases). All releases are tagged and released on GitHub.

Reviewer: Think we can remove this line here because we have it further down

Author: yeah good point!

Reviewer: Yeah thats fine

# Design System Library Deprecation and Release Policy

## Overview

> This policy outlines the procedures and timelines for deprecating components, utilities, and other features in the design system library. It also details the release process to ensure smooth transitions and maintain the stability and reliability of the library.

Reviewer 2: Given I'm suggesting to do release policy first (where deprecation is an aspect of our release policy and processes) I would switch up this overview to match

Author: Done :)

## Deprecation Policy

### Definition of Deprecation

> Deprecation is the process by which a component, utility, or feature is marked as obsolete and is scheduled for removal in a future release. Deprecation serves as a warning to users that they should transition to the recommended alternatives.

Reviewer 2: I'd put something along the lines of: 
"Deprecated content will usually have corresponding new content which provides an improved version of the deprecated functionality, however, on occasion functionality may be removed entirely. Reasoning and instructions for how to migrate from the deprecated to the new functionality (the migration path) will be provided alongside the notice of deprecation, where functionality is removed entirely, alternative implementations will be recommended."

Author: Added the below as per our chat : 

Reasoning and instructions for how to migrate from the deprecated to the new functionality (the migration path) will be provided alongside the notice of deprecation, where functionality is removed entirely, alternative implementations will be recommended.

Reviewer 3: Bit wordy and challenging to read for me... Just needs a full stop between deprecation and where i.e. deprecation. Where functionality is removed

Author: This is outdated now. Check the latest changes as the paragraph now look a bit different.

### Deprecation Guidelines

1. **Announcement**: Deprecation of a component, utility, or feature will be announced in the release notes of the version in which the deprecation first occurs. It will include:

    - The rationale for deprecation.

>     - Suggested alternatives or migration paths.

Reviewer 2: Small and not that important probs - but I'd put 'Migration paths or suggested alternatives' - as I think of suggested alternatives for when there's not a like-for-like functionality replacing the deprecated code, which is less common

Author: Done

    - The timeline for removal.

2. **Documentation**:

    - All deprecated items will be clearly marked in the documentation with a "Deprecated" label.
    - Documentation will include information on the deprecation and guidance on transitioning to alternatives.
    - Our documentation about the component is displayed on the [Service Manual](https://service-manual.ons.gov.uk/).

3. **Codebase**:

>     - Deprecated items will include a deprecation warning in the console, if applicable, to alert developers during development.

Reviewer: I dont think we should say this because most of our breaking changes will be nunjucks/html based so this wont be possible most of the time unless we specifically write some js to do this but I dont think thats a good idea either

2. **Release**:

    - **Release Notes**: Detailed release notes are published with each release, documenting new features, bug fixes, deprecations, and breaking changes.

>     - **Versioning**: The new version is tagged in the repository, and the library is published to the package manager (e.g., npm).

Reviewer: ```suggestion
    - **Versioning**: The new version is tagged in the repository, and the library is published to the package manager (e.g. npm).
```

>     - **Versioning**: The new version is tagged in the repository, and the library is published to the package manager (e.g., npm).

Reviewer: ```suggestion
    - **Versioning**: The new version is tagged in the repository and published, and the library is published to the package manager (e.g., npm).
```

>     - **Change Log**: Our change log is available on [GitHub](https://github.com/ONSdigital/design-system/releases). All releases are tagged and made through the release option of GitHub.

Reviewer: ```suggestion
    - **Change Log**: Our change log is available on [GitHub](https://github.com/ONSdigital/design-system/releases). All releases are tagged and released on GitHub.
```
Thought this read better

> 2. **Grace Period**: There will be at least two subsequent minor releases or a minimum of six months (whichever is longer) before the deprecated item is removed.

Reviewer: Does is matter if there is minimum 2 minor releases?

3. **Removal**: Deprecated items will be removed in the next major release following the grace period.

> ## Release Policy

Reviewer 2: I would put the release policy first, then deprecation info - so it follows a more 'general to specific' trajectory

Author: Done :)

### Versioning

We follow [Semantic Versioning](https://semver.org/) to manage releases:

-   **Major Releases (X.0.0)**: Introduce breaking changes. Deprecated items are removed, and significant new features or changes are added.
-   **Minor Releases (X.Y.0)**: Add new features and enhancements in a backward-compatible manner. May include deprecations.
-   **Patch Releases (X.Y.Z)**: Include backward-compatible bug fixes and documentation updates.

> ### Release Schedule

Reviewer 2: We should also have a schedule for minor releases
We could use the 'cadence' terminology here - to be in-keeping with the sector, but I kind of don't see the point of the term as it could cause confusion for those who don't know what it means. Then again, we could just define the term..

Author: This is something that we can add later once we agree on the cadence and process to use to do minor releases and patch releases.

    - The rationale for deprecation.
    - Migration paths or suggested alternatives.
    - The timeline for removal.

2. **Documentation**:

    - All deprecated items will be clearly marked in the documentation with a "Deprecated" label.
    - Documentation will include information on the deprecation and guidance on transitioning to alternatives.
    - Our documentation about the component is displayed on the [Service Manual](https://service-manual.ons.gov.uk/).

3. **Codebase**:

>     - Deprecated items will include a deprecation warning in the console, if applicable, to alert developers during development.

Reviewer: I dont think we should say this because most of our breaking changes will be nunjucks/html based so this wont be possible most of the time unless we specifically write some js to do this but I dont think thats a good idea either

    - Deprecated code will be maintained and supported (excluding new features) until its removal.

### Deprecation Timeline

1. **Initial Announcement**: Deprecated items will be announced and documented in a minor release.
2. **Grace Period**: There will be at least two subsequent minor releases or a minimum of six months (whichever is longer) before the deprecated item is removed.
3. **Removal**: Deprecated items will be removed in the next major release following the grace period.

## Communication

> -   **Changelog**: Maintain a changelog with detailed release notes for every release.

Reviewer: Wouldnt this just be our release notes on Github and we can link to that? Just thinking Id rather not duplicate this if we dont have to

-   **Deprecation Notices**: Send deprecation notices via appropriate channels (e.g., mailing lists, Slack, documentation site) to ensure all users are informed.
-   **Migration Guides**: Provide detailed migration guides and tools to assist users in transitioning from deprecated items to new alternatives.

> ## Review and Update

Reviewer 2: do we want a place for any suggestions to go? or do we not take suggestions on this haha

Author: I have added a link to the get in touch page of the service manual :)
